programs/mango-v4/src/instructions/token_deposit.rs

```diff
@@ -100,20 +100,38 @@ impl<'a, 'info> DepositCommon<'a, 'info> {
         &self,
         remaining_accounts: &[AccountInfo],
         amount: u64,
+        reduce_only: bool,
         allow_token_account_closure: bool,
     ) -> Result<()> {
         require_msg!(amount > 0, "deposit amount must be positive");
 
-        let token_index = self.bank.load()?.token_index;
+        let mut bank = self.bank.load_mut()?;
+        let token_index = bank.token_index;
+
+        let (amount_i80f48, token_index) = {
+            // Get the account's position for that token index
+            let account = self.account.load()?;
+            let position = account.token_position(token_index)?;
+
+            if bank.is_reduce_only() {
+                require!(reduce_only, MangoError::SomeError);
+            }
+            let amount_i80f48 = if reduce_only {
+                position
+                    .native(&bank)
+                    .min(I80F48::ZERO)
+                    .abs()
+                    .min(I80F48::from(amount))
+            } else {
+                I80F48::from(amount)
+            };
+            (amount_i80f48, token_index)
+        };
```

Comment:
need to take care here, amount_i80f48 is potentially 0

Reply:
q: why would 0 need special handling?

Reply:
it was just a reminder to myself; I think spl transfers of 0 amount will succeed